content/posts/2022-02-02-decreased-backend-tests-step/index.md

```diff
@@ -0,0 +1,173 @@
+
+
+## Methodology
+
+We followed a scientific method for problem-solving based on the following steps:
+
+- Identify a problem:
+We may have several potential issues in our code impacting the backend test step execution time.
+However, we should pick one to focus on. In this step, we took several problems and guessed which one of them would be better to start with.
+
+- Research:
+After having a clear definition of the problem, we started doing some research to understand how it was being handled by the software community.
+
+- Hypothesis:
+Based on the results of previous steps, we could gather some hypotheses in the format of "what if...".
+
+- Experimentation:
+For each hypothesis, we have evaluated the effort to test it. After qualifying it as a valid one, we prototyped it.
+We did it for a minimal set of changes enabling an initial evaluation. Our goal here was to change minimal and get the max result possible.
+
+- Results and conclusions:
+Once we had a prototype working as expected, we started gathering data. If it showed promising results, we could invest more effort in it.
+
+## Actions and Results
+
+### Baseline
+
+As a baseline, we considered the total execution time for the backend test step: 22 minutes.
+
+1. Map and split slowest Django apps
+
+We had a bunch of tests spread through several Django apps.
+
+We came up with the following hypothesis: What if we split the faster/slower apps into different jobs?
+
+We started by measuring the execution time for each Django app. Then, we created manually 2 test jobs: the first would have the slowest tests and the second would have the fastest. We tried to split it equally based on total execution time, ie, the second job with faster tests would have more tests than the first.
+
+We got the following results showing an improvement of about 7 minutes
+
+- Job 1 - 14 min 35 s
+- Job 2 - 14 min 15 s
+
+It showed a valid path which was splitting tests to run in parallel. However, breaking them into multiple jobs was not feasible, since we'd keep adding more and more tests daily.
+
+2. No migration before backend tests steps
+
+When we were measuring closer the tests execution time, we realized that the pipeline was taking about 8 minutes to run Django migrations (more than 1400 migration files) since we had a lot of databases and cross-references.
+
+Since we were using [pytest-django](https://pytest-django.readthedocs.io/en/latest/index.html), which takes the current model definition to be used as an in-memory database,
+we could remove the migrations steps.
+
+By removing them, we saw a bunch of tests failing. It turned out that we were using Django migration for data migration too. To fix it, we created several pytest fixtures to provide required data instead of inserting them into the database.
+
+By the end of this step, we could get an improvement of 8 minutes in the total pipeline execution time.
```

Comment:
What was the status at this point? How long was the pipeline taking?

Reply:
I've tried to get back our data in this point in time, but I couldn't 😢